### ISSUE TITLE
[Q] Does support for embedding and conv2d layers in bnb.py at tuners/lora make sense?

### ISSUE THREAD

**Issue by GM-git-dotcom** [huggingface/peft#1392]
### Feature request

Currently, `bnb.py` at `peft/tuners/lora` has classes for `Linear4bit` and `Linear8bit`. But `layers.py` at `peft/tuners/lora` defines implementations for LoRA in `Linear`, `Embedding`, and `Conv2d` (I believe in the future, it would also include `Conv3d` as per #1193).

I was wondering if it makes sense to extend quantization support for `Embedding` and `Conv2d` layers? Could they be a thing in `bnb.py`? (possibly in other tuners as well). Sorry if this is obvious.

### Motivation

I was just curious about `bnb.py` only having support for Linear when LoRA can be implemented through other layer types (please correct me if I'm wrong).

### Your contribution

With some help, I would be happy to contribute a PR towards this, starting with bnb for `Embedding`.

**Comment by BenjaminBossan** [comment]
Yes, in principle, it should be possible add support for the `Embedding` layer from bnb to PEFT. Feel free to open a PR and ping us so that we can give feedback and help with open questions.

Regarding `Conv2d`, I don't think it's supported by bnb and thus could not be easily added to PEFT.

**Comment by asr-sheep1** [comment]
Thank you very much for making such a great project open source, I learned a lot of new things by using PEFT, thank you again. I have a puzzle of  `class Linear8bitLt` in `peft/src/peft/tuners/lora/bnb.py`. By reading the code I understand that it is different from the definition in bitsanbytes because the latter uses int8 precision during calculation. I used PEFT in Alpaca-lora finetune. In the transfoemer model obtained with `is_load_8bit`, the parts (q, v) replaced by lora still use `Linear8bitLt `on the surface, but actually use `torch.nn.Linear` in `peft/src/peft/tuners/lora/layer.py update_layer()`.  Lora layer does not use int8 to participate in the calculation, it is for the sake of accuracy. Is that right, or did I not understand the whole calling program correctly?

**Comment by younesbelkada** [comment]
Hi @asr-sheep1 

Below is what happens under the hood if you initialize a `Linear8bitLt` class from peft

![lora_animated](https://github.com/huggingface/peft/assets/49240599/36bccc04-130c-482b-b3e8-870bf6cd4136)

In blue you have the original linear layer (which is in 8bit) and we simply attach the LoRA adapters next to it and calculate the sum of both hidden states.

The original `Linear8bit` layer performs some part of the matmul in 8-bit and other parts (outliers) in 16-bit but the resulting hidden states will be in 16-bit precision. See https://huggingface.co/blog/hf-bitsandbytes-integration for more details

**Comment by asr-sheep1** [comment]
> Hi @asr-sheep1
> 
> Below is what happens under the hood if you initialize a `Linear8bitLt` class from peft
> 
> ![lora_animated](https://private-user-images.githubusercontent.com/49240599/311656054-36bccc04-130c-482b-b3e8-870bf6cd4136.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxNTg0ODgsIm5iZiI6MTcxMDE1ODE4OCwicGF0aCI6Ii80OTI0MDU5OS8zMTE2NTYwNTQtMzZiY2NjMDQtMTMwYy00ODJiLWIzZTgtODcwYmY2Y2Q0MTM2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDExNTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4M2JmZGNmNGYxODkxYTE2Y2UwNzU5ODA2Y2MzOGJhZWFiY2ZiNDJkMjMyMzFkNmNlNTU2NDNhY2RmODNjYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MkvJtcPeFaNwhzxvMjWZofsLKkd7C5zFifrDJbYngI8) [ ![lora_animated](https://private-user-images.githubusercontent.com/49240599/311656054-36bccc04-130c-482b-b3e8-870bf6cd4136.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxNTg0ODgsIm5iZiI6MTcxMDE1ODE4OCwicGF0aCI6Ii80OTI0MDU5OS8zMTE2NTYwNTQtMzZiY2NjMDQtMTMwYy00ODJiLWIzZTgtODcwYmY2Y2Q0MTM2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDExNTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4M2JmZGNmNGYxODkxYTE2Y2UwNzU5ODA2Y2MzOGJhZWFiY2ZiNDJkMjMyMzFkNmNlNTU2NDNhY2RmODNjYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MkvJtcPeFaNwhzxvMjWZofsLKkd7C5zFifrDJbYngI8) ](https://private-user-images.githubusercontent.com/49240599/311656054-36bccc04-130c-482b-b3e8-870bf6cd4136.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxNTg0ODgsIm5iZiI6MTcxMDE1ODE4OCwicGF0aCI6Ii80OTI0MDU5OS8zMTE2NTYwNTQtMzZiY2NjMDQtMTMwYy00ODJiLWIzZTgtODcwYmY2Y2Q0MTM2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDExNTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4M2JmZGNmNGYxODkxYTE2Y2UwNzU5ODA2Y2MzOGJhZWFiY2ZiNDJkMjMyMzFkNmNlNTU2NDNhY2RmODNjYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MkvJtcPeFaNwhzxvMjWZofsLKkd7C5zFifrDJbYngI8) [ ](https://private-user-images.githubusercontent.com/49240599/311656054-36bccc04-130c-482b-b3e8-870bf6cd4136.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxNTg0ODgsIm5iZiI6MTcxMDE1ODE4OCwicGF0aCI6Ii80OTI0MDU5OS8zMTE2NTYwNTQtMzZiY2NjMDQtMTMwYy00ODJiLWIzZTgtODcwYmY2Y2Q0MTM2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDExNTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4M2JmZGNmNGYxODkxYTE2Y2UwNzU5ODA2Y2MzOGJhZWFiY2ZiNDJkMjMyMzFkNmNlNTU2NDNhY2RmODNjYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MkvJtcPeFaNwhzxvMjWZofsLKkd7C5zFifrDJbYngI8)
> 
> In blue you have the original linear layer (which is in 8bit) and we simply attach the LoRA adapters next to it and calculate the sum of both hidden states.
> 
> The original `Linear8bit` layer performs some part of the matmul in 8-bit and other parts (outliers) in 16-bit but the resulting hidden states will be in 16-bit precision. See https://huggingface.co/blog/hf-bitsandbytes-integration for more details

Thanks!  I have reviewed this material. So, during the calculation process, Lora_A and Lora_B, these two layers of networks, did not use int8 matmul, right?

**Comment by younesbelkada** [comment]
Correct yes, Lora_A and Lora_B do not use int8 matmul in this case
